data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/consumer/impl/UnorderedOffsetManager.java

```diff
@@ -110,62 +105,143 @@ public UnorderedOffsetManager(final KafkaConsumer<?, ?> consumer, final Consumer
+   */
+  private static class OffsetTracker {
+
+    private final static int ACKS_GARBAGE_SIZE_THRESHOLD = 128; // Meaning 8192 messages are on hold
+    private final static long[] POWS = new long[64];
+
+    static {
+      // Initialize POWS
+      for (int i = 0; i < 64; i++) {
+        long pow = 0;
+        for (int j = 0; j <= i; j++) {
+          pow |= 1L << j;
+        }
+        POWS[i] = pow;
+      }
+    }
+
+    private long lastAcked;
+    private long[] uncommitted;
+
+    private int greaterBlockIndex;
+    private int greaterBitIndexInGreaterBlock;
+
+    public OffsetTracker(long initialOffset) {
+      this.lastAcked = initialOffset;
+      this.uncommitted = new long[1];
+      this.greaterBlockIndex = -1;
+      this.greaterBitIndexInGreaterBlock = -1;
+    }
+
+    public void recordNewOffset(long offset) {
+      long diffWithLastCommittedOffset = offset - this.lastAcked - 1;
+      int blockIndex = blockIndex(diffWithLastCommittedOffset);
+      int bitIndex = (int) (diffWithLastCommittedOffset % 64); // That's obviously smaller than a long
+
+      checkAcksArraySize(blockIndex);
+
+      // Let's record this bit and update the greater indexes
+      this.uncommitted[blockIndex] |= 1L << bitIndex;
+      if (this.greaterBlockIndex < blockIndex) {
+        this.greaterBlockIndex = blockIndex;
+        this.greaterBitIndexInGreaterBlock = bitIndex;
+      } else if (this.greaterBlockIndex == blockIndex && this.greaterBitIndexInGreaterBlock < bitIndex) {
+        this.greaterBitIndexInGreaterBlock = bitIndex;
+      }
+    }
+
+    public boolean shouldCommit() {
+      // Let's check if we have all the bits to 1, except the last one
+      for (int b = 0; b < this.greaterBlockIndex; b++) {
+        if (this.uncommitted[b] != POWS[63]) {
+          return false;
+        }
+      }
+
+      return this.uncommitted[this.greaterBlockIndex] == POWS[this.greaterBitIndexInGreaterBlock];
+    }
+
+    public long uncommittedSize() {
+      return this.greaterBitIndexInGreaterBlock + 1 + (greaterBlockIndex * 64L);
+    }
+
+    public long offsetToCommit() {
+      return this.lastAcked + uncommittedSize() + 1;
+    }
+
+    public void reset(long committed) {
+      this.lastAcked = committed - 1;
+      this.greaterBlockIndex = -1;
+      this.greaterBitIndexInGreaterBlock = -1;
+
+      // Cleanup the acks array or overwrite it, depending on its size
+      if (this.uncommitted.length > ACKS_GARBAGE_SIZE_THRESHOLD) {
+        // No need to keep that big array
+        this.uncommitted = new long[1];
+      } else {
+        Arrays.fill(this.uncommitted, 0L);
+      }
+    }
+
+    private int blockIndex(long val) {
+      return (int) (val >> 6);
+    }
+
+    private void checkAcksArraySize(int blockIndex) {
+      if (this.uncommitted.length < blockIndex) {
+        this.uncommitted = Arrays.copyOf(this.uncommitted, (blockIndex + 1) * 2);
+      }
+    }
+
+  }
```

Comment:
Looks good to me, however, I spent too much time figuring out what's going on, mind adding some comments on the overall idea and an example/walkthrough?

Reply:
I actually made a drawing on my tablet to explain to myself how this thing works, I'll try to put some comments together tomorrow

Reply:
That should do it